### PR TITLE
Add Subnet Validator - Milestone 3

### LINKED ISSUE #871: Enable node to be a subnet validator after validator is bootstrapped (Milestone 3)
We should be able to call entire all the nodes in a cluster to be subnet validators: e.g. avalanche node addSubnetValidator <clusterName> <subnetID>

### PR DESCRIPTION
**Pull request by sukantoraymond**
Closes #871

### CODE REVIEW

File: cmd/nodecmd/join.go
Comment (felipemadero) [reply cmd/nodecmd/join.go:347]: if errr...

File: pkg/constants/constants.go
Comment (felipemadero) [inline pkg/constants/constants.go:114]: can we use status from avalanchego vms platformvm status ?

File: pkg/constants/constants.go
Comment (felipemadero) [inline pkg/constants/constants.go:115]: an we use status from avalanchego vms platformvm status ?

File: cmd/nodecmd/join.go
Comment (felipemadero) [inline cmd/nodecmd/join.go:362]: an we use status from avalanchego vms platformvm status ?

File: cmd/nodecmd/join.go
Comment (felipemadero) [inline cmd/nodecmd/join.go:364]: can we use status from avalanchego vms platformvm status ?

File: cmd/nodecmd/join.go
Comment (felipemadero) [inline cmd/nodecmd/join.go:325]: I will insist a little more on this clusterName stuff

File: cmd/nodecmd/join.go
Comment (felipemadero) [inline cmd/nodecmd/join.go:327]: is this needed? and the same for the other one

Comment (felipemadero) [inline cmd/nodecmd/join.go:327]: if err ...

File: cmd/nodecmd/join.go
Comment (felipemadero) [inline cmd/nodecmd/join.go:338]: if err

File: cmd/nodecmd/join.go
Comment (felipemadero) [inline cmd/nodecmd/join.go:358]: if err

File: cmd/nodecmd/join.go
Comment (felipemadero) [inline cmd/nodecmd/join.go:314]: if err

File: cmd/nodecmd/join.go
Comment (felipemadero) [inline cmd/nodecmd/join.go:321]: not needed?

Comment (felipemadero) [reply cmd/nodecmd/join.go:321]: I think it should print in other place. also, just return isBootstrapped

File: pkg/constants/constants.go
Comment (felipemadero) [inline pkg/constants/constants.go:81]: where are this values comming from?

File: pkg/constants/constants.go
Comment (felipemadero) [inline pkg/constants/constants.go:83]: can generat this two from the other two in code, or otherwise?

File: cmd/nodecmd/join.go
Comment (felipemadero) [inline cmd/nodecmd/join.go:229]: I would calculate the Str from the other one, or otherwise

File: cmd/nodecmd/join.go
Comment (felipemadero) [inline cmd/nodecmd/join.go:231]: same

File: cmd/nodecmd/join.go
Comment (felipemadero) [inline cmd/nodecmd/join.go:204]: you should also check after prompting if that is the case

File: cmd/nodecmd/join.go
Comment (felipemadero) [inline cmd/nodecmd/join.go:220]: constant for last arg

File: cmd/nodecmd/join.go
Comment (felipemadero) [inline cmd/nodecmd/join.go:221]: just return err

File: cmd/nodecmd/join.go
Comment (felipemadero) [inline cmd/nodecmd/join.go:117]: if err

File: cmd/nodecmd/join.go
Comment (felipemadero) [inline cmd/nodecmd/join.go:87]: maybe we can have one function for the block

File: cmd/nodecmd/join.go
Comment (felipemadero) [inline cmd/nodecmd/join.go:94]: if err

File: cmd/nodecmd/join.go
Comment (felipemadero) [inline cmd/nodecmd/join.go:72]: if err

**Review by felipemadero — COMMENTED** [review]
Pretty good stuff

**Comment by felipemadero** [comment]
> > Why automatic-cloud-deploy-existing-sg-kp is the base?
> 
> That was milestone 2 branch and we merged it

so you can change the base to main?

File: pkg/subnet/public.go
Comment (felipemadero) [reply pkg/subnet/public.go]: Please use "AddValidator transaction" as that is the label on the ledger device

File: pkg/subnet/public.go
Comment (felipemadero) [inline pkg/subnet/public.go:134]: Please use "Please sign AddValidator transaction on the ledger device" as that is the message in ledger I believe

File: pkg/prompts/validations.go
Comment (felipemadero) [inline pkg/prompts/validations.go:63]: pls update this also

File: pkg/prompts/validations.go
Comment (felipemadero) [inline pkg/prompts/validations.go:66]: please update this to the same value as if condition

File: pkg/constants/constants.go
Comment (felipemadero) [inline pkg/constants/constants.go:75]: I think this should be removed

File: pkg/constants/constants.go
Comment (felipemadero) [inline pkg/constants/constants.go:74]: this constant should be removed

File: pkg/application/app.go
Comment (felipemadero) [inline pkg/application/app.go:684]: probably name it RemoveAnsibleStatusDir, and not pass argument

File: pkg/application/app.go
Comment (felipemadero) [inline pkg/application/app.go:672]: probably name it CreateAnsibleStatusFile and only pass the filename not the full file path
or have one create per ansible status file kind

File: pkg/ansible/playbook/setupNode.yml
Comment (felipemadero) [reply pkg/ansible/playbook/setupNode.yml:15]: ok let's close conversation after that

File: pkg/ansible/playbook/setupNode.yml
Comment (felipemadero) [reply pkg/ansible/playbook/setupNode.yml:18]: Please dynamically create the file in some way, instead on relying it already exists.
Either make ansible to create the remote file with the expected contents (ignore on how to do that), or make cli
to create a tmp file and make ansible to copy it.

File: cmd/nodecmd/join.go
Comment (felipemadero) [inline cmd/nodecmd/join.go:38]: this also should be a cmdline flag

File: pkg/constants/constants.go
Comment (felipemadero) [inline pkg/constants/constants.go:83]: I would like @dasconnor to consider on leading time for validators and posible issues from that (if any)

File: cmd/nodecmd/join.go
Comment (felipemadero) [reply cmd/nodecmd/join.go:380]: when trying to add subnet validator, if first try to add primary validator, if already a primary, if exists with error. it should not

File: cmd/nodecmd/join.go
Comment (felipemadero) [inline cmd/nodecmd/join.go:142]: if err...

File: cmd/nodecmd/join.go
Comment (felipemadero) [inline cmd/nodecmd/join.go:313]: if err

File: cmd/nodecmd/join.go
Comment (felipemadero) [inline cmd/nodecmd/join.go:456]: if already primary validator, this will exit with error and not add subnet validator. it should not. it should continue and add.

File: cmd/nodecmd/join.go
Comment (felipemadero) [inline cmd/nodecmd/join.go:87]: I think not ok cases should return error. same for other instances of this pattern

File: cmd/nodecmd/join.go
Comment (felipemadero) [reply cmd/nodecmd/join.go:98]: I believe the code block is almost repeated for the three outputs, json, result, only change the field name and in one case it is a bool

**Review by dasconnor — COMMENTED** [review]
Started looking over this, need to a do a more in-depth review but I'm a bit too tired on a friday evening.

File: cmd/nodecmd/join.go
Comment (dasconnor) [inline cmd/nodecmd/join.go:231]: Ticket?

File: cmd/nodecmd/join.go
Comment (dasconnor) [inline cmd/nodecmd/join.go:398]: I thought on 229 the start time was 1 minute?

File: cmd/nodecmd/join.go
Comment (dasconnor) [inline cmd/nodecmd/join.go:466]: You should rename this to `convertNanoAvaxToAvaxString` or something like that. Name isn't clear.

File: pkg/constants/constants.go
Comment (dasconnor) [reply pkg/constants/constants.go:81]: Yeah, let's do 5 minutes for all

**Review by dasconnor — COMMENTED** [review]
Really nice job. Just a couple small things. If you can resolve all the conversations, I think we'll be good to merge.

File: pkg/ansible/playbook/setupNode.yml
Comment (felipemadero) [reply pkg/ansible/playbook/setupNode.yml:15]: https://github.com/ava-labs/avalanche-cli/issues/951

File: pkg/ansible/playbook/setupNode.yml
Comment (felipemadero) [inline pkg/ansible/playbook/setupNode.yml:19]: can we just do `shell:  echo '{"MetricsEnabled}":false}' > /home/ubuntu/.avalanche-cli/config` instead of copying a local file (that also may not be present or configured in the expected way)

File: cmd/nodecmd/join.go
Comment (felipemadero) [inline cmd/nodecmd/join.go:406]: if err ...

File: cmd/nodecmd/join.go
Comment (felipemadero) [inline cmd/nodecmd/join.go:405]: err is ignored, please make checkNodeIsPrimaryNetworkValidator to return (bool, error) to differentiate check failure from validtion status, then use the value to either return quickly cause it is already a validator, or aftertwards add it as validator

File: cmd/nodecmd/join.go
Comment (felipemadero) [reply cmd/nodecmd/join.go:418]: Yup I suggested so for ux (if not the user waits a minute here) but either case I just defer to @dasconnor opinion here for the different validator lead times

**Review by felipemadero — COMMENTED** [review]
I just have a couple more comments, afterwards LGTM

File: pkg/ansible/playbook/setupNode.yml
Comment (felipemadero) [reply pkg/ansible/playbook/setupNode.yml:19]: I see. Anyway I believe copying a file from local machine is not needed in this case, unless there is a reason to keep local configuration? Also there may be other configuration in the file not expected by the server. Would like to know @dasconnor opinion on this one

Comment (dasconnor) [reply pkg/ansible/playbook/setupNode.yml:19]: I'm ok copying the file, that way we might get some extra metrics haha

**Review by felipemadero — APPROVED** [review]

File: cmd/nodecmd/join.go
Comment (dasconnor) [reply cmd/nodecmd/join.go:421]: As long as it's consistent, it's fine

**Pull request merged**
